stream.go

```diff
@@ -0,0 +1,432 @@
+	//
+	// If there was an error, all future calls to Next will return ok == false.
+	Next() (stream string, entries []StreamEntry, ok bool)
+}
+
+// NewStreamReader returns a new StreamReader for the given client.
+//
+// Any changes on opts after calling NewStreamReader will have no effect.
+func NewStreamReader(c Client, opts StreamReaderOpts) StreamReader {
+	sr := &streamReader{c: c, opts: opts}
+
+	if sr.opts.Group != "" {
+		sr.cmd = "XREADGROUP"
+		sr.fixedArgs = []string{"GROUP", sr.opts.Group, sr.opts.Consumer}
+	} else {
+		sr.cmd = "XREAD"
+		sr.fixedArgs = nil
+	}
+
+	if sr.opts.Count > 0 {
+		sr.fixedArgs = append(sr.fixedArgs, "COUNT", strconv.Itoa(sr.opts.Count))
+	}
+
+	if !sr.opts.NoBlock {
+		dur := 5 * time.Second
+		if sr.opts.Block < 0 {
+			dur = 0
+		} else if sr.opts.Block > 0 {
+			dur = sr.opts.Block
+		}
+		msec := int(dur / time.Millisecond)
+		sr.fixedArgs = append(sr.fixedArgs, "BLOCK", strconv.Itoa(msec))
+	}
+
+	if sr.opts.Group != "" && sr.opts.NoAck {
+		sr.fixedArgs = append(sr.fixedArgs, "NOACK")
+	}
+
+	sr.streams = make([]string, 0, len(sr.opts.Streams))
+	sr.ids = make(map[string]string, len(sr.opts.Streams))
+	for stream, id := range sr.opts.Streams {
+		sr.streams = append(sr.streams, stream)
+
+		if id != nil {
+			sr.ids[stream] = id.String()
+		} else if sr.cmd == "XREAD" {
+			sr.ids[stream] = "$"
+		} else if sr.cmd == "XREADGROUP" {
+			sr.ids[stream] = ">"
+		}
+	}
+
+	// set to nil so we don't accidentally use it later, since the user could have changed
+	// the map after using the reader.
+	sr.opts.Streams = nil
+
+	sr.fixedArgs = append(sr.fixedArgs, "STREAMS")
+	sr.fixedArgs = append(sr.fixedArgs, sr.streams...)
+
+	// preallocate space for all arguments passed to Cmd
+	sr.args = make([]string, 0, len(sr.fixedArgs)+len(sr.streams))
+
+	return sr
+}
+
+// streamReader implements the StreamReader interface.
+type streamReader struct {
+	c    Client
+	opts StreamReaderOpts // copy of the options given to NewStreamReader with Streams == nil
+
+	streams []string
+	ids     map[string]string
+
+	cmd       string   // command. either XREAD or XREADGROUP
+	fixedArgs []string // fixed arguments that always come directly after the command
+	args      []string // arguments passed to Cmd. reused between calls to Next to avoid allocations.
+
+	unmarshaler streamReaderUnmarshaler
+	err         error
+}
+
+func (sr *streamReader) do() error {
+	sr.args = append(sr.args[:0], sr.fixedArgs...)
+
+	for _, s := range sr.streams {
+		sr.args = append(sr.args, sr.ids[s])
+	}
+
+	return sr.c.Do(Cmd(&sr.unmarshaler, sr.cmd, sr.args...))
+}
+
+// Err implements the StreamReader interface.
+func (sr *streamReader) Err() error {
+	return sr.err
+}
+
+// Next implements the StreamReader interface.
+func (sr *streamReader) Next() (stream string, entries []StreamEntry, ok bool) {
+	if sr.err != nil {
+		return "", nil, false
+	}
+
+	if len(sr.unmarshaler.A) == 0 {
+		if sr.err = sr.do(); sr.err != nil {
```

Comment:
I updated the documentation for now.

We could check for `Conn` inside `NewStreamReader` and wrap the `Conn` in `timeoutOk` like this:

```go
if conn, ok := c.(Conn); ok {
	c = timeoutOk(conn)
}
```

Then we could just ignore errors of type `timeoutError` (or better, check for a `Timeout` method like [net.Error](https://golang.org/pkg/net/#Error) has, which could be added to `timeoutError`) and make `Next()` return `ok == false` if `err.Timeout() == true` and than don't set the `sr.err` field so users could continue to call `Next`.

Reply:
Pushed a new commit that ignores temporary timeouts. I kept the comment on `Block` since it's still a good idea to make sure the commands don't timeout before the specified `Block` duration, but the StreamReader should now work after timeouts. I also updated the PubSub stuff to use the same logic.